### PR TITLE
Initial RequestRestartNow API

### PR DESCRIPTION
**Pull request by aeloros**
Create a new Restart API that restarts the calling process under the same context.  For packaged processes, the CoreApplication API is utilized.

### CODE REVIEW

File: dev/AppLifecycle/AppInstance.cpp [hunk 1]
```diff
@@ -333,6 +338,95 @@ namespace winrt::Microsoft::Windows::AppLifecycle::implementation
...
     }
 
+    std::wstring AppInstance::GenerateRestartAgentPath()
+    {
+        // Calculate the path to the restart agent as being in the same directory as the current module.
+        wil::unique_hmodule module;
+        THROW_IF_WIN32_BOOL_FALSE(GetModuleHandleEx(GET_MODULE_HANDLE_EX_FLAG_FROM_ADDRESS, reinterpret_cast<PCWSTR>(AppInstance::RequestRestartNow), &module));
+
+        auto modulePath = GetModulePath(module.get());
+        path fullPath = modulePath.substr(0, modulePath.find_last_of(L'\\'));
```
Comment (jaholme) [inline dev/AppLifecycle/AppInstance.cpp:348]: >ath fullPath = modulePath.substr(0, modulePath.find_last_of(L'\\')); [](http://example.com/codeflow?start=9&length=68)

use std::filesystem::path::parent_path

File: dev/AppLifecycle/Association.cpp [hunk 2]
```diff
@@ -42,7 +24,7 @@ namespace winrt::Microsoft::Windows::AppLifecycle::implementation
         do
         {
             path_size = static_cast<uint32_t>(path.size());
-            actual_size = ::GetModuleFileName(nullptr, path.data(), path_size);
```
Comment (jaholme) [inline dev/AppLifecycle/Association.cpp:45]: why not use wil::GetModuleFileNameW?

File: dev/AppLifecycle/AppInstance.cpp (hunk 1 above)
Comment (aeloros) [reply dev/AppLifecycle/AppInstance.cpp:348]: <3

File: dev/AppLifecycle/AppInstance.cpp [hunk 3]
```diff
@@ -333,6 +338,95 @@ namespace winrt::Microsoft::Windows::AppLifecycle::implementation
...
+        size_t handlesToInheritCount{ 1 };
+        HANDLE* handlesToInherit = reinterpret_cast<HANDLE*>(parentHandle.addressof());
+        THROW_IF_WIN32_BOOL_FALSE(UpdateProcThreadAttribute(attributeList.get(), 0, PROC_THREAD_ATTRIBUTE_HANDLE_LIST, handlesToInherit, 
+            handlesToInheritCount * sizeof(HANDLE), nullptr, nullptr));
+
+        STARTUPINFOEX info{};
+        info.StartupInfo.cb = sizeof(info);
+        info.lpAttributeList = attributeList.get();
+        
+        PROCESS_INFORMATION processInfo{};
```
Comment (jaholme) [inline dev/AppLifecycle/AppInstance.cpp:413]: wil::unique_process_info?

File: dev/RestartAgent/main.cpp [hunk 4]
```diff
@@ -0,0 +1,45 @@
...
+    std::wstring cmdLine{ GetCommandLineW() };
+    wil::unique_hlocal_ptr<PWSTR[]> argv{ CommandLineToArgvW(cmdLine.c_str(), &argc) };
+
+    wil::unique_handle parentHandle;
+    auto handleArg{ argv[1] };
+    parentHandle.reset(reinterpret_cast<HANDLE>(_wtoi64(handleArg)));
+
+    // Determine the exact path of the calling process needing a restart.
+    wchar_t callerPath[MAX_PATH];
+    DWORD size = _countof(callerPath);
```
Comment (jaholme) [inline dev/RestartAgent/main.cpp:26]: use wil::QueryFullProcessImageNameW so we don't introduce a max_path limitation

File: dev/AppLifecycle/AppInstance.cpp (hunk 3 above)
Comment (aeloros) [reply dev/AppLifecycle/AppInstance.cpp:413]: ?

File: dev/AppLifecycle/AppInstance.cpp [hunk 5]
```diff
@@ -333,6 +338,95 @@ namespace winrt::Microsoft::Windows::AppLifecycle::implementation
...
+        THROW_IF_WIN32_BOOL_FALSE(CreateProcess(exePath.c_str(), cmdLine.get(), nullptr, nullptr, TRUE, EXTENDED_STARTUPINFO_PRESENT, nullptr, nullptr, 
+            &info.StartupInfo, &processInfo));
+
+        // Close the thread handle immediately, but assign the process handle to an RAII object to handle cleanup for the failure paths.
+        ::CloseHandle(processInfo.hThread);
+        wil::unique_handle processHandle(processInfo.hProcess);
+
```
Comment (jaholme) [inline dev/AppLifecycle/AppInstance.cpp:420]: should we transfer foreground rights to the agent here?

Comment (aeloros) [reply dev/AppLifecycle/AppInstance.cpp:420]: Yes, good catch!

File: dev/AppLifecycle/AppInstance.cpp (hunk 3 above)
Comment (aeloros) [reply dev/AppLifecycle/AppInstance.cpp:413]: I think you mean unique_process_information.

**Review by jaholme — APPROVED** [review]
:shipit:

File: dev/AppLifecycle/Association.cpp (hunk 2 above)
Comment (aeloros) [reply dev/AppLifecycle/Association.cpp:45]: Great idea!

File: dev/AppLifecycle/AppInstance.cpp [hunk 6]
```diff
@@ -333,6 +338,109 @@ namespace winrt::Microsoft::Windows::AppLifecycle::implementation
...
+        {
+            return winrt::Windows::ApplicationModel::Core::CoreApplication::RequestRestartAsync(arguments).get();
+        }*/
+
+        // Only one restart can happen at a time.
+        auto mutexName = wil::str_printf<std::wstring>(L"%s_RequestRestartNowInProgress", ComputeAppId().c_str());
+        wil::unique_mutex restartMutex;
+
+        restartMutex.create(mutexName.c_str(), CREATE_MUTEX_INITIAL_OWNER, MUTEX_ALL_ACCESS);
+        if (GetLastError() == ERROR_ALREADY_EXISTS)
```
Comment (riverar) [inline dev/AppLifecycle/AppInstance.cpp:374]: `create` is not documented as having `SetLastError` semantics, so this is not reliable. Instead, I suggest checking the Boolean return from `create`.

File: dev/AppLifecycle/AppInstance.cpp [hunk 7]
```diff
@@ -333,6 +338,109 @@ namespace winrt::Microsoft::Windows::AppLifecycle::implementation
...
+            return AppRestartFailureReason::RestartPending;
+        }
+
+        auto releaseOnExit{ restartMutex.ReleaseMutex_scope_exit() };
+
+        // Use DuplicateHandle to get a real handle from the pseudo-handle returned by GetCurrentProcess().  A real handle
+        // is required in order for it to be inherited 
+        wil::unique_handle parentHandle;
+        THROW_IF_WIN32_BOOL_FALSE(DuplicateHandle(GetCurrentProcess(), GetCurrentProcess(), GetCurrentProcess(), wil::out_param(parentHandle), 
+            PROCESS_QUERY_INFORMATION | SYNCHRONIZE | PROCESS_TERMINATE, TRUE, 0));
```
Comment (riverar) [inline dev/AppLifecycle/AppInstance.cpp:385]: `PROCESS_QUERY_INFORMATION` may be too much here for protected or privileged processes. (Are those supported?) Consider perhaps `PROCESS_QUERY_LIMITED_INFORMATION` if possible.

File: dev/RestartAgent/main.cpp [hunk 8]
```diff
@@ -0,0 +1,69 @@
...
+
+    wil::unique_handle parentHandle;
+    auto handleArg{ argv[1] };
+    parentHandle.reset(reinterpret_cast<HANDLE>(_wtoi64(handleArg)));
+
+    // Determine the exact path of the calling process needing a restart.
+    auto callerPath = wil::QueryFullProcessImageNameW<wil::unique_cotaskmem_string>(parentHandle.get(), 0);
+
+    // Terminate the caller.  TerminateProcess is async, so we need to wait for the process to finish terminating.
+    THROW_IF_WIN32_BOOL_FALSE(TerminateProcess(parentHandle.get(), 0));
```
Comment (riverar) [inline dev/RestartAgent/main.cpp:28]: App requesting restart is never given a chance to spin down, which could result in leaks (e.g. GDI handles). I noted the same concern in the design document PR.

File: dev/AppLifecycle/AppInstance.cpp (hunk 7 above)
Comment (aeloros) [reply dev/AppLifecycle/AppInstance.cpp]: Just saw this.  I think I fixed this already (probably after you mentioned it though).  Sorry I'm a little slow catching up after the holidays!

File: dev/AppLifecycle/AppInstance.cpp (hunk 6 above)
Comment (aeloros) [reply dev/AppLifecycle/AppInstance.cpp:375]: This code is correct I believe.

File: dev/AppLifecycle/AppInstance.cpp [hunk 9]
```diff
@@ -358,13 +358,14 @@ namespace winrt::Microsoft::Windows::AppLifecycle::implementation
             featureUsageReported = true;
         }
 
-        // TODO: Redirect UWP requests to CoreApplication::RequestRestartAsync().
-
-        // For packaged, redirect to the already existing API.
-        /*if (IsPackagedProcess())
+        // If a better way of detecting UWP is created in the future, this check should change.
+        if (IsPackagedProcess() && wil::get_token_is_app_container())
```
Comment (NoUWP) [inline dev/AppLifecycle/AppInstance.cpp:362]: AppContainer and Packaged Process are not UWP specific.  https://scorpiosoftware.net/2019/01/15/fun-with-appcontainers/

Whereas `CoreWindow` is.

**Pull request merged**